source/OrderedTypes/SupLattice.lagda

```diff
@@ -0,0 +1,286 @@
+open import Slice.Family
+open import UF.ImageAndSurjection pt
+
+open AllCombinators pt fe
+open PropositionalTruncation pt
+
+\end{code}
+
+We commence by defining sup lattices and some boiler plate. 
+
+\begin{code}
+
+module sup-lattice-def (𝓤 𝓦 𝓥 : Universe) where
+
+ sup-lattice-data : 𝓤  ̇ → 𝓤 ⊔ 𝓦 ⁺ ⊔ 𝓥 ⁺  ̇
+ sup-lattice-data A = (A → A → Ω 𝓦) × (Fam 𝓥 A → A)
+ 
+ is-sup-lattice : {A : 𝓤  ̇} → sup-lattice-data A → 𝓤 ⊔ 𝓦 ⊔ 𝓥 ⁺  ̇  
+ is-sup-lattice {A} (_≤_ , ⋁_) = is-partial-order A _≤_ × suprema
+  where
+   open Joins _≤_
+   suprema : 𝓤 ⊔ 𝓦 ⊔ 𝓥 ⁺  ̇
+   suprema = (U : Fam 𝓥 A) → ((⋁ U) is-lub-of U) holds
+
+ sup-lattice-structure : 𝓤 ̇ → 𝓤 ⊔ 𝓥 ⁺ ⊔ 𝓦 ⁺ ̇
+ sup-lattice-structure A = Σ d ꞉ (sup-lattice-data A) , is-sup-lattice d
+
+Sup-Lattice : (𝓤 𝓦 𝓥 : Universe) → (𝓤 ⊔ 𝓦 ⊔ 𝓥) ⁺  ̇
+Sup-Lattice 𝓤 𝓦 𝓥 = Σ A ꞉ 𝓤  ̇ , sup-lattice-structure A
+ where
+  open sup-lattice-def 𝓤 𝓦 𝓥
+
+⟨_⟩ : Sup-Lattice 𝓤 𝓦 𝓥 → 𝓤  ̇
+⟨ A , rest ⟩ = A
+
+order-of : (L : Sup-Lattice 𝓤 𝓦 𝓥) → (⟨ L ⟩ → ⟨ L ⟩ → Ω 𝓦)
+order-of (A , (_≤_ , ⋁_) , rest) = _≤_
+
+syntax order-of L x y = x ≤⟨ L ⟩ y
+
+join-of : (L : Sup-Lattice 𝓤 𝓦 𝓥) → Fam 𝓥 ⟨ L ⟩ → ⟨ L ⟩
+join-of (A , (_≤_ , ⋁_) , rest) = ⋁_
+
+syntax join-of L U = ⋁⟨ L ⟩ U
+
+partial-orderedness-of : (L : Sup-Lattice 𝓤 𝓦 𝓥)
+                       → is-partial-order ⟨ L ⟩ (order-of L)
+partial-orderedness-of (A , (_≤_ , ⋁_) , order , is-lub-of) = order
+
+reflexivity-of : (L : Sup-Lattice 𝓤 𝓦 𝓥) → is-reflexive (order-of L) holds
+reflexivity-of L = pr₁ (pr₁ (partial-orderedness-of L))
+
+antisymmetry-of : (L : Sup-Lattice 𝓤 𝓦 𝓥) → is-antisymmetric (order-of L) 
+antisymmetry-of L = pr₂ (partial-orderedness-of L)
+
+transitivity-of : (L : Sup-Lattice 𝓤 𝓦 𝓥) → is-transitive (order-of L) holds
+transitivity-of L = pr₂ (pr₁ (partial-orderedness-of L))
+
+join-is-lub-of : (L : Sup-Lattice 𝓤 𝓦 𝓥)
+               → (U : Fam 𝓥 ⟨ L ⟩)
+               → ((order-of L) Joins.is-lub-of join-of L U) U holds
+join-is-lub-of (A , (_≤_ , ⋁_) , order , suprema) = suprema
+
+join-is-upper-bound-of : (L : Sup-Lattice 𝓤 𝓦 𝓥)
+                       → (U : Fam 𝓥 ⟨ L ⟩)
+                       → ((order-of L) Joins.is-an-upper-bound-of
+                                          join-of L U) U holds
+join-is-upper-bound-of L U = pr₁ (join-is-lub-of L U)
+
+join-is-least-upper-bound-of : (L : Sup-Lattice 𝓤 𝓦 𝓥)
+                             → (U : Fam 𝓥 ⟨ L ⟩)
+                             → ((u' , _) : Joins.upper-bound (order-of L) U)
+                             → (order-of L (join-of L U) u') holds
+join-is-least-upper-bound-of L U = pr₂ (join-is-lub-of L U)
+
+sethood-of : (L : Sup-Lattice 𝓤 𝓦 𝓥) → is-set ⟨ L ⟩
+sethood-of L =
+  type-with-prop-valued-refl-antisym-rel-is-set
+   (λ x → λ y → order-of L x y holds)
+   (λ x → λ y → holds-is-prop (order-of L x y))
+   (λ x → reflexivity-of L x)
+   (λ x → λ y → antisymmetry-of L)
+
+\end{code}
+
+We now define monotone endomaps on a sup-lattice. This notion is not too
```

Comment:
Since this will now become a more general purpose file, I would recommend defining `is-monotone` for maps between two (potentially) different sup-lattices and then defining `is-monotone-endomap` as a special case.